File: src/aoc_2021/day2.clj
(ns aoc-2021.day2
  (:require [clojure.string :as str]))

(defn simple-nav
  [acc dir n]
  (cond
    (= dir "forward") (update acc :hp #(+ % n))
    (= dir "up") (update acc :depth #(- % n))
    (= dir "down") (update acc :depth #(+ % n))))

(defn aiming-nav
  [acc dir n]
     (cond
       (= dir "forward") (-> acc
                             (update :hp #(+ % n))
                             (update :depth #(+ % (* (:aim acc) n))))
       (= dir "up") (update acc :aim #(- % n))
       (= dir "down") (update acc :aim #(+ % n))))

(defn navigate
  [course f acc]
  (reduce
   (fn [acc line]
     (let [[dir sn] (str/split line #" ")
           n (Integer/parseInt sn)]
       (f acc dir n)))
   acc
   course))

(defn navigate-simple
  [course]
  (navigate course simple-nav {:hp 0 :depth 0}))

(defn navigate-with-aim
  [course]
  (navigate course aiming-nav {:hp 0 :depth 0 :aim 0}))

File: src/day2.rs
pub fn navigate(course: &str) -> (i32, i32) {
    course.lines().fold((0, 0), |(hp, d), line| {
        let (dir, n) = line.split_once(" ").unwrap();
        match (dir, n.parse::<i32>().unwrap()) {
            ("forward", n) => (hp + n, d),
            ("up", n) => (hp, d - n),
            ("down", n) => (hp, d + n),
            _ => unreachable!(),
        }
    })
}

pub fn navigate_with_aim(course: &str) -> (i32, i32, i32) {
    course.lines().fold((0, 0, 0), |(hp, d, aim), line| {
        let (dir, n) = line.split_once(" ").unwrap();
        match (dir, n.parse::<i32>().unwrap()) {
            ("forward", n) => (hp + n, d + (aim * n), aim),
            ("up", n) => (hp, d, aim - n),
            ("down", n) => (hp, d, aim + n),
            _ => unreachable!(),
        }
    })
}
